File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // fastptr and slow ptr
        ListNode* prev = nullptr;
        ListNode* slowptr = head;
        ListNode* fastptr = head;
        for (int i = 0; i < n; i++) {
            fastptr = fastptr->next;
        }
        while (fastptr != nullptr) {
            prev = slowptr;
            slowptr = slowptr->next;
            fastptr = fastptr->next;
        }
        if (prev == nullptr) {
            return slowptr->next;
        } else {
            prev->next = slowptr->next;
        }
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # general idea: two pointers, one will move as per normal the other one will
        # move when it nth step after the first one
        
        # when len of linkedlist is 1, then it can only remove itself
        if not head.next:
            return head.next
        

        fast = head
        slow = head
        prev = None
        count = 0

        while fast:
            fast = fast.next
            if (count >= n):
                prev = slow
                slow = slow.next
            count += 1
        
        # end of loop: slow will be pointing to the node that is supposed to be removed
        # and prev will the node before slow

        if not prev:
            head = head.next
        else:
            prev.next = slow.next
        return head
        
        
        